src/main/java/org/mifos/connector/mpesa/camel/routes/SafaricomRoutesBuilder.java

```diff
@@ -194,12 +221,46 @@ public void configure() {
                     if(resultCode.equals("0")) {
                         exchange.setProperty(TRANSACTION_FAILED, false);
                     } else {
+                        exchange.setProperty(ERROR_CODE, resultCode);
                         exchange.setProperty(TRANSACTION_FAILED, true);
                     }
 
                     exchange.setProperty(SERVER_TRANSACTION_ID, server_id);
                     exchange.setProperty(TRANSACTION_ID, correlationId);
                 })
+                .choice()
+                .when(exchangeProperty(ERROR_CODE).isNotNull())
+                .to("direct:filter-by-error-code")
+                .process(exchange -> {
+                    Boolean isRecoverableError = exchange.getProperty(IS_ERROR_RECOVERABLE, Boolean.class);
+                    if (isRecoverableError) {
+                        exchange.setProperty(IS_TRANSACTION_PENDING, true);
+                    } else {
+                        exchange.setProperty(TRANSACTION_FAILED, true);
+                    }
+                })
+                .process(collectionResponseProcessor)
+                .otherwise()
+                .process(collectionResponseProcessor)
+                .when(header("CamelHttpResponseCode").isEqualTo("500"))
+                .process(exchange -> {
+                    JSONObject jsonObject = new JSONObject(exchange.getIn().getBody(String.class));
+                    String errorCode = jsonObject.getString("errorCode");
+                    exchange.setProperty(ERROR_CODE, errorCode);
```

Comment:
If error code is just the code like 5001 or DS026, where are we setting the English error description?

Reply:
With the current PR implementation. We are picking the errorCode/ resultCode and if it is not the success one then, we are getting the error code and seeing if it is recoverable or not and then making decision.

Reply:
and errorCode is a string not a int/long/double.